File: 03-provider-registry/golang/providers/registry.go
package providers

import (
	"context"
	"fmt"
	"strings"
)

var registry = map[string]Provider{}

func Register(p Provider) {
	registry[p.Name()] = p
}

// ChatWithFallback tries each provider in chain order; moves to the next on any error.
func ChatWithFallback(ctx context.Context, messages []Message, chain []string, assemble func([]Message, Provider) ([]Message, error)) (string, error) {
	var errs []string

	for _, name := range chain {
		p, ok := registry[name]
		if !ok {
			errs = append(errs, name+": not registered")
			continue
		}

		assembled, err := assemble(messages, p)
		if err != nil {
			errs = append(errs, fmt.Sprintf("%s (context assembly): %v", name, err))
			continue
		}

		reply, err := p.Chat(ctx, assembled)
		if err != nil {
			fmt.Printf("[provider:%s] failed — %v\n", name, err)
			errs = append(errs, fmt.Sprintf("%s: %v", name, err))
			continue
		}
		return reply, nil
	}

	return "", fmt.Errorf("all providers failed:\n%s", strings.Join(errs, "\n"))
}

File: 03-provider-registry/nodejs/src/providers/registry.ts
import type { Message, Provider } from './types.ts';
import { assembleContext } from '../context.ts';

// ── Provider registry ────────────────────────────────────────────────────────

const providerRegistry = new Map<string, Provider>();

export function registerProvider(provider: Provider) {
  providerRegistry.set(provider.name, provider);
}

// ── Fallback router ──────────────────────────────────────────────────────────

// Tries each provider in order; moves to the next on any error.
export async function chatWithFallback(
  messages: Message[],
  chain: string[],
): Promise<string> {
  const errors: string[] = [];

  for (const name of chain) {
    const provider = providerRegistry.get(name);
    if (!provider) {
      errors.push(`${name}: not registered`);
      continue;
    }

    const ctx = await assembleContext(messages, provider);
    try {
      return await provider.chat(ctx);
    } catch (err: any) {
      const msg = err?.message ?? String(err);
      console.warn(`[provider:${name}] failed — ${msg}`);
      errors.push(`${name}: ${msg}`);
    }
  }

  throw new Error(`All providers failed:\n${errors.join('\n')}`);
}
